### ISSUE TITLE
Missing unreachable code/arm warning, when a match guard is diverging

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust#134734]
I tried this code:

```rust
fn main() {
    match () {
        _ if loop {} => (),
        _ => (),
    }
    println!("nya :3");
}
```
([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=26c036b8ccf18d04209b39f2f623dafb))

I expected to see this happen: compiler to emit a warning that the first arm and later code is unreachable (since `loop{}` never terminates and has type `!`).

Instead, this happened: the code compiles without warnings.

### Meta

rustc version: ` 1.84.0-beta.4 (2024-12-07 202008a1b8de96d2e5b6)`

**Comment by compiler-errors** [comment]
This seems somewhat nontrivial to lint, since the match only unconditionally diverges due to the match guard in this way if the pattern is unconditional and there are no other patterns that could've matched prior. Seems like a pretty special case to have to hard-code into the HIR divergence analysis which is already a complete mess.
